src/main/java/net/sf/mzmine/modules/peaklistmethods/filtering/duplicatefilter/DuplicateFilterTask.java

```diff
@@ -137,92 +139,175 @@ public void run() {
-                .contains(secondRow.getAverageMZ());
-
-            // Compare rt
-            final boolean sameRT = rtTolerance.getToleranceRange(firstRow.getAverageRT())
-                .contains(secondRow.getAverageRT());
+            boolean sameMZRT = filterByAvgRTMZ ? // average or single feature
+                checkSameAverageRTMZ(firstRow, secondRow, mzTolerance, rtTolerance)
+                : checkSameSingleFeatureRTMZ(rawFiles, firstRow, secondRow, mzTolerance,
+                    rtTolerance);
 
             // Duplicate peaks?
-            if (sameID && sameMZ && sameRT) {
-
+            if (sameID && sameMZRT) {
+              // create consensus row in new filter
+              if (!mode.equals(FilterMode.OLD_AVERAGE)) {
+                // copy all detected features of row2 into row1
+                // to exchange gap-filled against detected features
+                createConsensusRow(rawFiles, firstRow, secondRow);
+              }
+              // second row deleted
+              n++;
               peakListRows[secondRowIndex] = null;
             }
           }
         }
+        // add to new list
+        newPeakList.addRow(firstRow);
       }
-
       processedRows++;
     }
 
-    // Create the new peak list.
-    final PeakList newPeakList =
-        new SimplePeakList(origPeakList + " " + suffix, origPeakList.getRawDataFiles());
-
-    // Add all remaining rows to a new peak list.
-    for (int i = 0; !isCanceled() && i < rowCount; i++) {
-
-      final PeakListRow row = peakListRows[i];
-
-      if (row != null) {
-
-        // Copy the peak list row.
-        final PeakListRow newRow = new SimplePeakListRow(row.getID());
-        PeakUtils.copyPeakListRowProperties(row, newRow);
-
-        // Copy the peaks.
-        for (final Feature peak : row.getPeaks()) {
-
-          final Feature newPeak = new SimpleFeature(peak);
-          PeakUtils.copyPeakProperties(peak, newPeak);
-          newRow.addPeak(peak.getDataFile(), newPeak);
-        }
-
-        newPeakList.addRow(newRow);
-      }
-    }
-
+    // finalize
     if (!isCanceled()) {
-
       // Load previous applied methods.
       for (final PeakListAppliedMethod method : origPeakList.getAppliedMethods()) {
-
         newPeakList.addDescriptionOfAppliedTask(method);
       }
 
       // Add task description to peakList
       newPeakList.addDescriptionOfAppliedTask(
           new SimplePeakListAppliedMethod("Duplicate peak list rows filter", parameters));
+      LOG.info("Removed " + n + " duplicate rows");
     }
 
     return newPeakList;
   }
+
+  /**
+   * Turns firstRow to consensus row. With all features with highest FeatureStatus:
+   * DETECTED>ESTIMATED>UNKNOWN Or the highest feature when comparing two ESTIMATED features
+   * 
+   * @param rawFiles
+   * @param firstRow
+   * @param secondRow
+   */
+  private void createConsensusRow(RawDataFile[] rawFiles, PeakListRow firstRow,

```

Comment:
I think this method should return a new PeakListRow, if you are creating a consensus of two rows. If you simply keep using the firstRow, then the firstRow might be referenced from both the old and the new peak lists, which would create data integrity issues, no?

Reply:
I have already copied the firstRow to keep the original peak list unchanged. This method only exchanges some of the features. As mentioned in the comments. The first row is always copied and inserted. 
You want me to change it?